File: plugins/collector/event/apps/web/apache.go
package web

import (
	"collector/cache/process"
	"collector/event/apps"
	"regexp"
	"strings"
)

type Apache2 struct {
	version string
	reg     *regexp.Regexp
}

func (Apache2) Name() string { return "apache2" }

func (Apache2) Type() string { return "web" }

func (a Apache2) Version() string { return a.version }

func (Apache2) Match(p *process.Process) bool { return p.Name == "apache2" }

func (a *Apache2) Run(p *process.Process) (m map[string]string, err error) {
	if a.reg == nil {
		a.reg = regexp.MustCompile(`Apache2\/(\d+\.)+\d+`)
	}
	result, err := apps.Execute(p, "-v")
	str := a.reg.FindString(result)
	if str == "" {
		err = apps.ErrVersionNotFound
		return
	}
	a.version = strings.TrimPrefix(str, "Apache2/")
	return
}

func init() {
	apps.Regist(&Apache2{})
}

File: plugins/collector/src/event/apps/web/apache.rs
use crate::event::apps::{execute, find_version, AppProc, IApp};
use std::collections::HashMap;

pub struct Apache2 {
    version: String,
}

impl Apache2 {
    pub fn new() -> Self { Apache2 { version: String::new() } }
}

impl IApp for Apache2 {
    fn name(&self)     -> &'static str { "apache2" }
    fn app_type(&self) -> &'static str { "web" }
    fn version(&self)  -> &str         { &self.version }

    fn matches(&mut self, p: &AppProc) -> bool { p.name == "apache2" }

    fn run(&mut self, p: &AppProc) -> anyhow::Result<HashMap<String, String>> {
        let out = execute(p, &["-v"])?;
        // "Server version: Apache/2.4.54 (Debian)"
        self.version = find_version(&out);
        if self.version.is_empty() {
            return Err(anyhow::anyhow!("version not found"));
        }
        Ok(HashMap::new())
    }
}
